chain-signatures/node/src/logs.rs

```diff
@@ -61,22 +148,95 @@ where
-    let base_subscriber = Registry::default().with(EnvFilter::from_default_env());
-
-    if let Some(node_id) = node_id {
-        let fmt_layer =
-            tracing_subscriber::fmt::layer().event_format(NodeIdFormatter::new(node_id));
-        let subscriber = base_subscriber.with(fmt_layer);
-        tracing::subscriber::set_global_default(subscriber).expect("Failed to set subscriber");
-    } else if is_running_on_gcp() {
-        let stackdriver = stackdriver_layer().with_writer(std::io::stderr);
-        let subscriber = base_subscriber.with(stackdriver);
-        tracing::subscriber::set_global_default(subscriber).expect("Failed to set subscriber");
+fn get_resource(env: &str, node_id: &str) -> Resource {
+    static RESOURCE: OnceLock<Resource> = OnceLock::new();
+    RESOURCE
+        .get_or_init(|| {
+            Resource::builder()
+                .with_service_name(format!("mpc:{}:{}", env, node_id))
+                .with_attributes(vec![
+                    KeyValue::new("env", env.to_string()),
+                    KeyValue::new("node_id", node_id.to_string()),
+                ])
+                .build()
+        })
+        .clone()
+}
+
+async fn init_otlp_logs(env: &str, node_id: &str, otlp_endpoint: &str) -> SdkLoggerProvider {
+    let exporter = LogExporter::builder()
+        .with_http()
+        .with_endpoint(format!("{otlp_endpoint}/v1/logs"))
+        .build()
+        .expect("Failed to create log exporter");
+
+    SdkLoggerProvider::builder()
+        .with_batch_exporter(exporter)
+        .with_resource(get_resource(env, node_id))
+        .build()
+}
+
+async fn init_otlp_traces(env: &str, node_id: &str, otlp_endpoint: &str) -> SdkTracerProvider {
+    let exporter = SpanExporter::builder()
+        .with_http()
+        .with_endpoint(format!("{otlp_endpoint}/v1/traces"))
+        .build()
+        .expect("Failed to create trace exporter");
+
+    SdkTracerProvider::builder()
+        .with_sampler(Sampler::ParentBased(Box::new(Sampler::TraceIdRatioBased(
+            1.0,
+        ))))
+        .with_id_generator(RandomIdGenerator::default())
+        .with_resource(get_resource(env, node_id))
+        .with_batch_exporter(exporter)
+        .build()
+}
+
+pub async fn setup(env: &str, node_id: &str, options: &Options) -> OtlpGuard {
+    let log_otlp_provider = init_otlp_logs(env, node_id, options.otlp_endpoint.as_str()).await;
+    let log_otlp_layer = OpenTelemetryTracingBridge::new(&log_otlp_provider);
+
+    let log_fmt_layer = tracing_subscriber::fmt::layer()
+        .with_ansi(atty::is(atty::Stream::Stderr))
+        .with_line_number(true)
+        .with_thread_names(true)
+        .event_format(NodeIdFormatter::new(node_id))
+        .with_filter(EnvFilter::from_default_env());
+
+    let tracer_otlp_provider = init_otlp_traces(env, node_id, options.otlp_endpoint.as_str()).await;
+    let tracer_otlp = tracer_otlp_provider.tracer("mpc");
+
+    if is_running_on_gcp().await {
```

Comment:
since debug id is no longer specified from integration tests, can you add in a flag in Options that is specified by integration tests to disable GCP logging. Some of us are using GCP instances for development so integration tests should disable GCP related logging.

Reply:
It is a layered system. Presence of the GCP layer should not affect your experience.

Reply:
It does because it's a layered. All logs lose their colored formatting and are captured within `{}` when outputted within a GCP environment. GCP log manager will parse it but not so much for our terminals which is a terrible experience for development on top of GCP VMs